src/cursor/cur_version.c

```diff
@@ -0,0 +1,387 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __curversion_set_key --
+ *     WT_CURSOR->set_key implementation for version cursors.
+ */
+static void
+__curversion_set_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    table_cursor->set_key(table_cursor);
+}
+
+/*
+ * __curversion_get_key --
+ *     WT_CURSOR->get_key implementation for version cursors.
+ */
+static int
+__curversion_get_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    va_list ap;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    WT_ERR(__wt_cursor_get_keyv(cursor, cursor->flags, ap));
+    WT_ERR(__wt_cursor_get_keyv(table_cursor, table_cursor->flags, ap));
+
+err:
+    va_end(ap);
+    return (ret);
+}
+
+/*
+ * __curversion_next --
+ *     WT_CURSOR->next method for version cursors. The next function will position the cursor on the
+ *     next update of the key it is positioned at. We traverse through updates on the update chain,
+ *     then the ondisk value, and finally from the history store.
+ */
+static int
+__curversion_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_ITEM *hs_key, *hs_value;
+    WT_SESSION_IMPL *session;
+    WT_TIME_WINDOW tw;
+    WT_UPDATE *upd;
+    wt_timestamp_t ondisk_durable_stop_ts;
+    wt_timestamp_t ondisk_stop_ts;
+    uint64_t hs_counter, ondisk_stop_txn;
+    uint32_t hs_btree_id, hs_upd_type;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+
+    /* The cursor should be positioned. */
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_INT))
+        goto err;
+
+    upd = version_cursor->next_upd;
```

Comment:
The new upd is not saved to version_cursor->next_upd. The time next is called we start from the start again, which is wrong.